### DISCUSSION TITLE
Accessing db from go without using as a framework

### DISCUSSION THREAD

**Discussion by n-ham** [pocketbase/pocketbase#2683] [Q&A] [answered]
Is it possible to read from and write to the database when doing a go backend that just uses normal https routing from net/http? Otherwise are there some more examples of eg. doing post requests when using PB as a framework? As a novice when it comes to databases, would love some more basic examples of how to do some basic querying and record additions..

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand your use case, but if you want to access the db file from some 3rd party application - Yes, kind of (you can also check https://www.sqlite.org/faq.html#q5), but I wouldn't recommend it unless you really need it because the realtime events and app hooks may not work as expected since they are handled on app level.

The recommended approach to interact with the database is either [using PocketBase as Go framework](https://pocketbase.io/docs/use-as-framework/) OR via the exposed [Web APIs](https://pocketbase.io/docs/api-records/).

> Otherwise are there some more examples of eg. doing post requests when using PB as a framework?

I'm confused. Do you use PocketBase as Go framework or interact with it via the Web APIs?

When using PocketBase as framework, you can check https://pocketbase.io/docs/record-methods/.

When interacting with the Web APIs, you can find documentation in the _"Admin UI > {:yourCollection} > API Preview"_ section or in the general docs located in https://pocketbase.io/docs/api-records/.

If the above is not what you are looking for, please elaborate and/or provide an example of your use case.

**Comment by n-ham** [comment]
>I'm confused. Do you use PocketBase as Go framework or interact with it via the Web APIs?

I'm still trying to work this out myself a bit. I figured it'd be better to keep business logic on the backend and have the server do most of the db work and send anything needed to the client. I think in order to do that one might be best using PB as a framework rather than wanting to continue using vanilla net/http server..

Would love a 'post' example at https://pocketbase.io/docs/use-as-framework/ showing how to read form parameters, and for handling file uploads with 'post' requests as well.. 

Thanks for the very useful information..

**Reply by ganigeorgiev** [reply, marked as answer]
Internally PocketBase uses [echo v5_alpha](https://github.com/labstack/echo) as a router.

The documentation will be updated with the next release, but here is an example registering a custom POST route:

```go
package main

import (
	"log"
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/forms"
	"github.com/pocketbase/pocketbase/models"
	"github.com/pocketbase/pocketbase/tools/filesystem"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		// POST /example
		e.Router.POST(
			"/example",
			func(c echo.Context) error {
				collection, err := app.Dao().FindCollectionByNameOrId("example")
				if err != nil {
					return apis.NewNotFoundError("Failed to fetch example collection.", err)
				}

				record := models.NewRecord(collection)

				// ---
				// Bind manually the request data
				// https://pkg.go.dev/github.com/labstack/echo/v5#Context
				// ---
				data := struct {
					Name      string `form:"name"`
					SomeField string `form:"someField"`
				}{}
				if err := c.Bind(&data); err != nil {
					return apis.NewBadRequestError("Failed to read request data.", err)
				}

				// do some data validations...

				record.Set("name", data.Name)
				record.Set("someField", data.SomeField)

				if err := app.Dao().SaveRecord(record); err != nil {
					return apis.NewBadRequestError("Failed to create the record.", err)
				}
				// ...

				// ---
				// OR you can also leave the request data loading and validations to forms.RecordUpsert
				// https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.5/forms#RecordUpsert
				// ---

				form := forms.NewRecordUpsert(app, record)

				if err := form.LoadRequest(c.Request(), ""); err != nil {
					return apis.NewBadRequestError("Failed to load the submitted data.", err)
				}

				// multipart/form-data files are automatically loaded
				// but you also have the option manually upload files from the filesystem
				f1, _ := filesystem.NewFileFromPath("/path/to/file1")
				f2, _ := filesystem.NewFileFromPath("/path/to/file2")
				form.AddFiles("yourFileField1", f1, f2)

				if err := form.Submit(); err != nil {
					return apis.NewBadRequestError("Failed to create the record.", err)
				}

				return c.JSON(http.StatusOK, record)
			},
			// optionally register middlewares
			// https://pocketbase.io/docs/router/#middlewares
			apis.ActivityLogger(app),
			apis.RequireAdminAuth(),
		)

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Comment by n-ham** [comment]
I have just been looking into the feasibility of using PB as a framework.. Is there an easy way to add a directory for serving static files? And how do we hook up certs for https support? 

I had a bit of a play with dark mode for the dashboard in the repo the other day too. It might not be a bad idea to tackle dark mode sooner rather than later, there were a few key things which didn't use css variables at all to change at the moment, and some things were very difficult to get a dark mode working with conflicting dark/dark or light/light bg/color combinations when tweaking existing css variables. It might help clean some of that up to tackle it sooner rather than later..

**Reply by ganigeorgiev** [reply]
> Is there an easy way to add a directory for serving static files? 

Yes, you can check for example the prebuilt executable [main.go file](https://github.com/pocketbase/pocketbase/blob/master/examples/base/main.go):

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    // serves static files from the provided public dir (if exists)
    e.Router.GET("/*", apis.StaticDirectoryHandler(os.DirFS("./pb_public"), indexFallback))
    return nil
})
```

> And how do we hook up certs for https support?

PocketBase comes with automanaged Let's Encrypt certificate when started with the `--https=yourdomain.com:443` flag. With the next release there will be an option to attach a custom issued certificate when using as framework (_it is in the [`develop` branch](https://github.com/pocketbase/pocketbase/tree/develop) via the new `OnBeforeSeve` `e.ServeEvent.Server` and `e.ServeEvent.CertManager` fields_).

**Reply by n-ham** [reply]
I'm having lots of problems with autocert, eg: `2023/06/14 01:22:53 http: TLS handshake error from *.*.*.*:52796: acme/autocert: missing server name`, I wonder if it'd be possible to trouble you for an example of attaching custom issued certs re what's mentioned here? Would be awesome to have CLI option for using custom generated certs as well. I'm hoping to see if I can get things working a little more consistently as I'm trying to automate everything for a BaaS platform I'm making called Cloudfort.. Sorry for being a bother, PocketBase is awesome though! 

I could potentially look into contributing to the docs with some of the stuff you've been helping me with if you'd be interested in docs contributions? 

Also https://go-acme.github.io/lego/ looks quite promising if you've never seen it, I'm currently using certbot otherwise (which seems to work more consistently than autocert) but like that lego waits for the dns records to propagate, useful when automating provisioning/initialisation etc..

**Comment by n-ham** [comment]
It would be awesome to be able to use PB with a go backend server that doesn't use PB as a framework, ie. be able to make the pocketbase app and be able to interact with the DB without having to do routes, serve static html/css/js files and do certs etc. through PB.. I wonder if it's possible to add a command other than serve for that?

**Reply by ganigeorgiev** [reply]
I'm sorry, I'm failing to understand what exactly do you mean. Could you provide a pseudo code or example with the desired command and how do you plan to use it?

As noted previously, if you want to just access the database from an external application, you can do that by operating directly with the `pb_data/data.db` SQLite file, but I don't understand how do you plan to interact with PocketBase and why do you need it in the first place if you are going with this route.

**Reply by ganigeorgiev** [reply]
>  be able to make the pocketbase app and be able to interact with the DB without having to do routes...

I think you can do that already by calling `Bootstrap()` instead of `Start()`, eg.:

```go
app := pocketbase.New()

app.Bootstrap()

// the Dao and the database is now accessible...

record, err := app.Dao().FindRecordById("example", "RECORD_ID")
```

But note that the Admin UI and any other web APIs (including the realtime subscriptions) will not be initialized and accessible.

**Reply by n-ham** [reply]
app.Boostrap() is what I was looking for, thanks! Pocketbase is pretty awesome! I've made this sweet little web based dash people can use with a terminal, file browser, editor and server monitor page, might even add in db management from that as well.. Gonna build a full stack platform for deploying on linode/vultr servers using PB, I'll show it off when it's done :)

**Comment by n-ham** [comment]
I'm wondering if it's possible to (add functionality to) pass an email and password to pocketbase to automate creating the first admin account?

**Reply by ganigeorgiev** [reply]
With v0.16.0 there is helper `admin` command:

```sh
# use --help for all options
./pocketbase admin create test@example.com 1234567890
```

If you don't use `Start()`, you can register it manually:

```go
app.RootCmd.AddCommand(cmd.NewAdminCommand(app))
```

Alternatively, you can also create the admin inside a [Migration](https://pocketbase.io/docs/migrations/):

```go
package migrations

import (
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/daos"
	m "github.com/pocketbase/pocketbase/migrations"
	"github.com/pocketbase/pocketbase/models"
)

func init() {
	m.Register(func(db dbx.Builder) error {
		dao := daos.New(db)

		admin := new(models.Admin)
		admin.Email = "test@example.com"
		admin.SetPassword("123456789")

		return dao.SaveAdmin(admin)
	}, func(db dbx.Builder) error {
		return nil
	})
}
```

**Reply by n-ham** [reply]
Perfect!

**Comment by n-ham** [comment]
When I do `pocketbase serve --http="yourdomain.com:80"` I am able to load the admin dashboard without any ssl, however I'm having no luck with `pocketbase serve --https="yourdomain.com:443"`, I am wondering if maybe I should be waiting longer for the certificate to kick in? Otherwise I'm not sure why it would work for http but not https? 

Where I have pocketbase running it's saying: 

`TLS handshake error from my-local-ip:26594: acme/autocert: missing certificate`

**Reply by ganigeorgiev** [reply]
I'm not sure what could be the cause of the error. If the domain is new, it may need some time to propagate the dns records. Note that Let's encrypt requires a valid public domain name in order to work (bare IPs for example are not supported).

**Reply by hungcrush** [reply]
@n-ham consider use nginx proxy 😬

**Reply by n-ham** [reply]
I've now got dns servers pointed at Linode so I can use their api to add/delete A records for project sub domains. Works a treat with the certs generated by certbot for serving my dash over https, but having trouble getting the certs issued through pocketbase to work consistently.. I get all sorts of messages like :

```
TLS handshake error from *.*.*.*:26594: acme/autocert: missing certificate
TLS handshake error from *.*.*.*:52796: acme/autocert: missing server name
TLS handshake error from *.*.*.*:39631: acme/autocert: host "matrix.c0mpl3x.xyz" not configured in HostWhitelist
```

Sorry for my ignorance, but how would nginx proxy help? I've not used it before..

**Comment by n-ham** [comment]
I am finding that if I first do `pocketbase serve --https="yourdomain.com:443"` and load the dashboard, then `pocketbase serve --https="yourdomain.com:8443"` then I can access via port 8443 (or even 4443 if I want) with ssl, but it will not work until I do it with port 443 first and load the page. Any ideas on how to avoid that?

**Reply by ganigeorgiev** [reply]
I'm sorry, I don't understand what do you mean. Please elaborate what you are trying to do and provide more details about your deployment environment and its configuration.

**Reply by ganigeorgiev** [reply]
Also please note that when performing certificate validations it is recommended `80` port also to be open, aka. you should have both `--http` and `--https` flags (see https://letsencrypt.org/docs/allow-port-80/).

Also, while technically it is allowed to have https on different port, as far as I know the validation can happen only on 80 and 443 (aka. you may need redirect from the other ports for the http-01 challenges).

**Reply by n-ham** [reply]
I am trying to serve over https using port 8443, but seems I need to first serve over port 443 load the page then serve over 8443. I have just found a solution from an automation pov but it's not ideal.. Basically I ssh into the machine, download and update pocketbase, then am finding for port 8443 to work I first need to serve port 443 and curl the page which I'd prefer to avoid if possible (though guess it's not the end of the world now that it works)

At the moment I have the following:

```
sshpass -p "$ROOT_PW" ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -l "root" $IP "mkdir pb; cd pb; wget -q -O pb.zip https://github.com/pocketbase/pocketbase/releases/download/v0.16.0/pocketbase_0.16.0_linux_amd64.zip; apt -y install unzip; unzip pb.zip; rm pb.zip; chmod a+x pocketbase; ./pocketbase update; ./pocketbase serve --https='$DOMAIN:443' & ./pocketbase admin create $ADMIN_EMAIL $ADMIN_PW; curl https://$DOMAIN/_/; killall pocketbase; ./pocketbase serve --https='$DOMAIN:8443' & disown; exit" &
```

**Reply by ganigeorgiev** [reply]
As mentioned in the above comment, certificate validations must happen on 80 or 443 port because of the let's encrypt http-01 challenges. Once the certificate is issued for your domain, you can technically use it with any other port as long as it is valid.

**Reply by ganigeorgiev** [reply]
Found a similar issue in the Let's encrypt forum - https://community.letsencrypt.org/t/port-4434-instead-of-443/61349 that confirms the above:

> https://community.letsencrypt.org/t/port-4434-instead-of-443/61349/4
> You can use a certificate on any port.
> Validation can only be performed over ports 80 and 443.

**Reply by n-ham** [reply]
Looks like I'll keep using the curl approach, thanks :)
